src/main/java/org/veupathdb/service/eda/ss/model/reducer/ancestor/DescendantCollapser.java

```diff
@@ -0,0 +1,78 @@
+package org.veupathdb.service.eda.ss.model.reducer.ancestor;
+
+import org.veupathdb.service.eda.ss.model.reducer.FilteredValueFile;
+import org.veupathdb.service.eda.ss.model.variable.VariableValueIdPair;
+import org.veupathdb.service.eda.ss.model.variable.binary.AncestorDeserializer;
+import org.veupathdb.service.eda.ss.model.variable.binary.ValueWithIdDeserializer;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Iterator;
+import java.util.function.Function;
+
+/**
+ * Given a sorted stream of entities and a mapping of those entities to their ancestors, returns a stream of ancestors
+ * where each ancestor entity return has at least one descendant in the stream of descendant entities.
+ */
+public class DescendantCollapser implements Iterator<Long> {
+  private final Iterator<VariableValueIdPair<Long>> ancestorStream;
+  private final Iterator<Long> currentEntityStream;
+  private VariableValueIdPair<Long> currentAncestor;
+  private Long currentEntity;
+  private Long lastAncestor;
+  private boolean hasStarted = false;
+
+  public DescendantCollapser(Path ancestorFilePath,
+                             AncestorDeserializer deserializer,
+                             Iterator<Long> currentEntityStream) throws IOException {
+    this.ancestorStream = new FilteredValueFile(ancestorFilePath,
+        x -> true,
+        new ValueWithIdDeserializer<>(deserializer),
+        Function.identity());
+    this.currentEntityStream = currentEntityStream;
+  }
+
+  public DescendantCollapser(Iterator<VariableValueIdPair<Long>> ancestorStream,
+                             Iterator<Long> currentEntityStream) {
+    this.ancestorStream = ancestorStream;
+    this.currentEntityStream = currentEntityStream;
+  }
+
+  @Override
+  public boolean hasNext() {
+    setCurrentIfNotStarted();
+    return this.currentAncestor != null && this.currentEntity != null;
+  }
+
+  @Override
+  public Long next() {
+    setCurrentIfNotStarted();
+    return nextMatch();
+  }
+
+  private void setCurrentIfNotStarted() {
+    if (!hasStarted) {
+      currentAncestor = ancestorStream.hasNext() ? ancestorStream.next() : null;
+      currentEntity = currentEntityStream.hasNext() ?  currentEntityStream.next() : null;
+      hasStarted = true;
+    }
+  }
+
+  private Long nextMatch() {
```

Comment:
i think we need the same approach here as in the Expander, with the method that advances whichever stream is behind the other

Reply:
I stared at this yesterday evening and managed to convince myself it would work. Fixed this morning and added the appropriate test case.